contrib/go/src/python/pants/contrib/go/tasks/go_buildgen.py

```diff
@@ -401,66 +405,57 @@ def generate_targets(self, local_go_targets=None):
-                "root, found:\n\t{}".format("\n\t".join(sorted(local_roots)))
+                "BUILD gen requires a single local Go source root, but found Go targets under "
+                f"multiple build roots: {', '.join(sorted(local_roots))}"
             )
         local_root = local_roots.pop()
 
-        if local_go_targets:
-            unrooted_locals = {t for t in local_go_targets if t.target_base != local_root}
-            if unrooted_locals:
-                raise self.UnrootedLocalSourceError(
-                    "Cannot BUILD gen until the following targets are "
-                    "relocated to the source root at {}:\n\t{}".format(
-                        local_root,
-                        "\n\t".join(sorted(t.address.reference() for t in unrooted_locals)),
-                    )
+        def infer_remote_root():
+            inferrable_roots = ["3rdparty/go", "3rd_party/go", "thirdparty/go", "third_party/go"]
+            msg = (
+                f"You can provide an explicit remote source root by setting remote_root in the "
+                f"[{self.options_scope}] section of your config file."
+            )
+            rr = None
+            for path in inferrable_roots:
+                if os.path.isdir(path):
+                    if rr:
+                        raise self.InvalidRemoteRootsError(
+                            f"Couldn't infer a single remote root. Found {rr} and {path}. {msg}"
+                        )
+                    rr = path
+            if not rr:
+                raise self.InvalidRemoteRootsError(
+                    f"Couldn't infer remote root.  Found none of: {','.join(inferrable_roots)}. {msg}"
                 )
-        else:
-            root = os.path.join(get_buildroot(), local_root)
-            local_go_targets = self.context.scan(root=root).targets(self.is_local_src)
-            if not local_go_targets:
-                return None
 
-        remote_roots = go_roots_by_category[SourceRootCategories.THIRDPARTY]
-        if len(remote_roots) > 1:
-            raise self.InvalidRemoteRootsError(
-                "Can only BUILD gen for a single Go remote library source "
-                "root, found:\n\t{}".format("\n\t".join(sorted(remote_roots)))
-            )
-        remote_root = remote_roots.pop() if remote_roots else None
+        remote_root_func = lambda: self.get_options().remote_root or infer_remote_root()
 
         generator = GoTargetGenerator(
             self.import_oracle,
             self.context.build_graph,
             local_root,
             self.get_fetcher_factory(),
             generate_remotes=self.get_options().remote,
-            remote_root=remote_root,
+            remote_root_func=remote_root_func,
         )
         with self.context.new_workunit("go.buildgen", labels=[WorkUnitLabel.MULTITOOL]):
             try:
                 generated = generator.generate(local_go_targets)
```

Comment:
It would be ideal if this returned `(generated, remote_root)`, but this works.

Reply:
Done, although that also required some contortion.